src/goTestExplorer.ts

```diff
@@ -157,4 +159,29 @@ export class GoTestExplorer {
         return symbols[0];
     }
 
+    private onSaveDocument(document: vscode.TextDocument) {
+        if (document.languageId !== 'go') {
+            return;
+        }
+        let testOnSave = vscode.workspace.getConfiguration('go')['testOnSave'];
+        if (!!testOnSave) {
+            this.goTestProvider.discoveredTests.
+                filter(t => t.isTestSuite).
+                filter(u => this.isSameGoTest(document.uri, u.uri)).
```

Comment:
@kroppt Nice work!!
we can use [getDiscoveredTestNode](https://github.com/ppparihar/GoTestExplorer/blob/master/src/goTestProvider.ts#L60) to get the expected `TestSuite` by the `uri` and it will help us to improve the performance. I added this method recently hence you will have to take the latest pull to use it.

People may start expecting added/removed tests to be reflect in the test explorer after this PR and we will have to take care of this.

Reply:
Thanks for the tip on performance. However, the test discoverer doesn't currently appear flexible enough to handle an updating of one test. If we discover all tests again, then they will not have been run yet, and we would need to run them all again, even though we only saved one file. Any ideas on how to how to improve the flexibility of the interface between the discoverer and the provider?

Reply:
[getTestFunctions](https://github.com/ppparihar/GoTestExplorer/blob/master/src/lib/testUtil.ts#L60-L70) can help you to get the all the tests which belongs to a test file. The same method used at [discoverTests](https://github.com/ppparihar/GoTestExplorer/blob/master/src/testDiscovery.ts#L33) which discover all the tests by tests files.

Next part is to update [discoveredTests](https://github.com/ppparihar/GoTestExplorer/blob/master/src/goTestProvider.ts#L13-L14). Currently it is happening at [onDicoveredTest](https://github.com/ppparihar/GoTestExplorer/blob/master/src/goTestProvider.ts#L77-L94) but it refresh all the tests and we want to refresh impacted tests only. we can pass a flag to [onDicoveredTest](https://github.com/ppparihar/GoTestExplorer/blob/master/src/goTestProvider.ts#L77-L94) and tweak the code to fulfill our need or we can create a new method to refresh impacted tests only. I'll prefer second approach. 

feel free to write back if you need more details.